c10/core/DispatchKeySet.cpp

```diff
@@ -46,8 +46,8 @@ bool isBackendDispatchKey(DispatchKey t) {
 // math_dispatch_keyset contains all keys in backend_dispatch_keyset and
 // autograd_dispatch_keyset Alias key DispatchKey::CompositeImplicitAutograd
 // maps to [math_dispatch_keyset x full_backend_mask]
-constexpr DispatchKeySet math_dispatch_keyset =
-    backend_dispatch_keyset | autograd_dispatch_keyset;
+constexpr DispatchKeySet math_dispatch_keyset = backend_dispatch_keyset |
+    autograd_dispatch_keyset | DispatchKeySet(DispatchKey::Python);
```

Comment:
This probably deserves a comment for why we're putting the (backend) `Python` key into `math_dispatch_keyset`, and not `backend_dispatch_keyset`?

My understanding is basically: if we did the latter, that would cause all `CompositeExplicitAutograd` kernels to decompose before python subclasses, which... would be pretty BC-breaking at least. Putting it in `math_dispatch_keyset` preserves the current state better, while also fixing some bugs (like preventing re-entrant python dispatch from running `CompositeImplicitAutograd` decomps).

Reply:
So I actually have no idea. I tried the other version and stuff segfaulted, and so I reverted back to this.

Intuitively, I think the idea is we only want to apply "math" decomps automatically for Python key; the more low level backend explicit autograd decomps should be opt in. But in principle it seems to me that CompositeExplicitAutograd should be ok too, it just isn't.

Reply:
I tried debugging through the PR where we had Python decompose at the CompositeExplicitAutograd level. Some of the segfaults looked like TensorIterator was getting passed a torch_dispatch Tensor Subclass, which is not what we want. So there might just be a mechanism problem somewhere